### ISSUE TITLE
[Bug Report] ttnn.transformer.scaled_dot_product_attention: the tile padding of K and V reaches the output when the sequence length is not a multiple of 32 — silently, PCC ~0

### ISSUE THREAD

**Issue by ayewo** [tenstorrent/tt-metal#57608]
### Component / Area

ops

### Issue Type (optional)

Bad Outputs

### Observed

`ttnn.transformer.scaled_dot_product_attention` (non-causal, no `attn_mask`) masks the key positions past the logical sequence length itself, so what K and V hold in their implicit tile padding should not affect the result. It does, whenever the sequence length is not a multiple of 32.

With large values in the padding rows of both K and V the output is wrong: PCC −0.003 against torch, max abs error 5.5e36. With the same values in K's padding alone it is degraded: PCC 0.973, RMSE 8× the clean case. V's padding alone has no effect, and at a length that is a multiple of 32 there is no padding and every case matches the clean one. Nothing raises, and with finite padding the output is finite too.

The same bfloat16 tensors, padding included, through `ttnn.matmul` → `ttnn.softmax` → `ttnn.matmul` on device are unaffected in every case, so this is specific to SDPA rather than to the precision.

`[1, 1, T, 64]`, bfloat16, default compute kernel config, padding value `1e38` (finite in bfloat16). PCC, max abs error and RMSE are against torch on the logical values:

| T | K padding | V padding | SDPA PCC | SDPA max abs err | SDPA RMSE | matmul-softmax-matmul PCC | its RMSE |
|---:|---|---|---:|---:|---:|---:|---:|
| 449 | 0 | 0 | 0.99966 | 0.0904 | 0.00897 | 0.99955 | 0.0114 |
| 449 | 1e38 | 0 | 0.97271 | 1.25 | 0.0716 | 0.99954 | 0.0115 |
| 449 | 0 | 1e38 | 0.99966 | 0.0904 | 0.00897 | 0.99955 | 0.0114 |
| 449 | 1e38 | 1e38 | −0.00298 | 5.48e36 | 1.37e36 | 0.99954 | 0.0115 |
| 450 | 0 | 0 | 0.99967 | 0.116 | 0.00887 | 0.99958 | 0.0115 |
| 450 | 1e38 | 0 | 0.97259 | 1.23 | 0.0721 | 0.99958 | 0.0115 |
| 450 | 0 | 1e38 | 0.99967 | 0.116 | 0.00887 | 0.99958 | 0.0115 |
| 450 | 1e38 | 1e38 | −0.00303 | 5.65e36 | 1.41e36 | 0.99958 | 0.0115 |
| 448 | any | any | 0.99966 | 0.0937 | 0.00897 | 0.99956 | 0.0114 |

A second script (below) sweeps lengths, padding values and operands at B=2 with 8 heads, under `HiFi4` and `fp32_dest_acc_en=True`, which also moves SDPA from its streaming compute kernel to its standard one. There the K-only case matches the clean one, but K and V together still give PCC ~0 at every length tried that is not a multiple of 32 (385, 417, 433, 449, 450, 451, 465, 481, 897) and are clean at 448. +Inf, −Inf and NaN padding behave like `1e38`: the output stays finite and is just as wrong.

### Expected

The result does not depend on what the tile padding holds: every row at PCC ≥ 0.999, as the explicit chain on the same tensors already is. If SDPA needs zeroed padding in K and V, it should zero it or say so, rather than return a wrong result.

### Ruled out by measurement

| hypothesis | result |
|---|---|
| bfloat16 precision | `matmul → softmax → matmul` on the same bf16 tensors, padding included, is unaffected in every case |
| non-finite values | finite `1e38` is enough; +Inf, −Inf and NaN behave the same, and the output stays finite |
| Q's padding | large values in Q's padding, alone or with one of K and V, are harmless (HiFi4, fp32 accumulation) |
| one compute kernel | `fp32_dest_acc_en=True` selects the standard kernel instead of the streaming one; the K-only case is then clean, but K and V together still give PCC ~0 |
| the model's tensors | random tensors with planted padding reproduce it; the model's captured q, k and v give PCC 0.0116, and 0.99999 once their padding is zeroed |

### How it was found

CosyVoice-300M's flow-matching decoder (a UNet with transformer blocks) produced a mel spectrogram about half non-finite for one Japanese utterance on Blackhole, and the vocoder output railed at its clamp. Bisecting inward: the attention path (the explicit chain is clean, SDPA is not) → the first failing SDPA call, at T = 449, whose K and V padding held 131 and 104 non-finite values → `ttnn.conv1d`, which writes large values into its output padding (up to 1.5e37 from an input whose padding is zero, and Inf/NaN at some lengths). LayerNorm leaves those rows huge, and the QKV projection overflows them into Inf/NaN in K and V together.

Across the ten lengths we measured, the model failed exactly when a UNet level's length was 1 mod 32, so tests at a fixed geometry never reach it.

### 1. Steps (exact commands)

Two files, `torch` and `ttnn` only, no model and no checkpoint — both pasted in full below.

```bash
python3 repro_sdpa_kv_padding.py                                                         # default compute config
python3 probe_sdpa_padding.py --values big,inf,-inf,nan --operands q,k,v,kv,qk,qv,qkv   # HiFi4, fp32 accumulation
```

### 2. Input data / link or description

All inputs generated in-script, deterministic, `torch.manual_seed(0)`, no external files.

- q `~ 0.3 · N(0, 1)`, k and v `~ N(0, 1)`, each `[1, 1, T, 64]`, bfloat16, `TILE_LAYOUT`, DRAM interleaved.
- The padding value is planted by uploading `[1, 1, ceil32(T), 64]` and restoring the logical shape with `ttnn.reshape(t, logical, padded, skip_padding_fill=True)`.
- `is_causal=False`, `scale=1.0`, no `attn_mask`, no program config (so `exp_approx_mode` takes its default, true), default compute kernel config, `ttnn.open_device(device_id=0)`.
- Reference: `torch.softmax(q @ k.T, -1) @ v` on the logical values, in float32.
- The sweep: q, k, v `[2, 8, T, 64]` seeded by T, `HiFi4` with `fp32_dest_acc_en=True`, reference in float64; a case is wrong when PCC ≤ 0.999 or any output is non-finite.

### 3. Frequency

always — deterministic, at every length tried that is not a multiple of 32.

### 1. Software Versions

`ayewo/tt-metal` at `571b1e0395`, which merges upstream `main` at `fda1e45f96f` (2026-08-12) and makes no changes under `ttnn/cpp/ttnn/operations/transformer/`.

Re-run on 2026-09-23 (UTC) on a second p150a, with both scripts below: every printed figure is identical at `571b1e0395`, at `ayewo/tt-metal` `f09a4992dab` (upstream base `201315d14c2`, 2026-09-16), and at upstream `main` `e249306523f` (2026-09-23). That `main` includes #57180 (accurate exponential mode), which does not apply here: without a program config, `exp_approx_mode` stays true.

### 2. Hardware Details

- Blackhole p150a, two boards, firmware bundle 18.10.0 — fails on both, with identical figures
- Wormhole — not tested

### Is this a regression?

Unknown.

### Impact

Silent wrong output from a supported op whenever a producer leaves large or non-finite values in the tile padding of K and V, which ops are free to do. In our model the producer is `ttnn.conv1d`, upstream of LayerNorm and the QKV projection.

We are not blocked. Zero-filling the padding of K and V with `ttnn.fill_implicit_tile_padding(t, 0.0)` before the call restores the clean result. It works in place, so it is safe inside a trace, and it costs 1.8 % of our traced flow stage, against 55 % for switching to the explicit chain.

### Suggested direction (take or leave)

Both kernels add the padded-key mask to the scores rather than writing it. With the default config (`fp32_dest_acc_en=False`) SDPA runs its streaming kernel, which L1-accumulates a mask tile onto QKᵀ for the last, partial K tile: 0 on valid columns, bf16 −inf (`0xFF80`) on padded ones (`generate_lightweight_mask_tiles` and `fill_vertical_tile_bf16` in `dataflow_common.hpp`). With `fp32_dest_acc_en=True` the standard kernel does `add_block_inplace(cb_qk_im, cb_mask_in, …)` with a Bfp4_b mask. The measurements say the addition does not always leave a padded column with zero weight: V's padding reaches the output only when K's padding is large, i.e. when the padded scores themselves are large. Writing −inf into the padded score positions instead of adding it, or excluding V's rows past the logical length from P·V, would make the result independent of the padding.

Happy to run variants of this on Blackhole.

### Reproducer

<details>
<summary><code>repro_sdpa_kv_padding.py</code></summary>

```python
# SPDX-FileCopyrightText: (c) 2026 Tenstorrent AI ULC
#
# SPDX-License-Identifier: Apache-2.0
"""ttnn.transformer.scaled_dot_product_attention is not isolated from the tile padding of K and V.

Non-causal, no attn_mask, sequence length not a multiple of 32. SDPA generates its own
padding mask (-inf on padded key columns), so the contents of K's and V's implicit tile
padding should not matter. They do: when both hold large values the output is wrong
(PCC ~0 against torch), and large values in K's padding alone degrade it. Zero padding,
large values in V's padding alone, and a length that is a multiple of 32 are unaffected.

The values used here are finite (1e38 is representable in bfloat16), so this is not a
NaN-propagation effect. Each case also runs the same tensors through matmul -> softmax ->
matmul on device, as a same-precision control. Model-free; B=1, one head, head dim 64.

    python repro_sdpa_kv_padding.py
"""
import torch

import ttnn


def pcc(a, b):
    a, b = a.flatten().double(), b.flatten().double()
    a, b = a - a.mean(), b - b.mean()
    return float((a * b).sum() / (a.norm() * b.norm()))


def err(o, ref):
    """PCC, max abs error and RMSE against the torch reference."""
    d = (o.double() - ref.double())
    return f"PCC {pcc(o, ref):+.5f} max|err| {float(d.abs().max()):<9.3g} RMSE {float(d.pow(2).mean().sqrt()):<9.3g}"


def upload(x, T, pad_value, device):
    """[1, 1, T, 64] logical, with `pad_value` written into rows T..Tpad of the tile padding."""
    Tp = -(-T // 32) * 32
    p = torch.full((1, 1, Tp, 64), pad_value)
    p[..., :T, :] = x
    t = ttnn.from_torch(p, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
    # A view with the logical length restored; skip_padding_fill keeps pad_value in place.
    return ttnn.reshape(t, ttnn.Shape([1, 1, T, 64]), ttnn.Shape([1, 1, Tp, 64]), skip_padding_fill=True)


def main():
    import os
    import subprocess

    home = os.environ.get("TT_METAL_HOME", ".")
    rev = subprocess.run(["git", "-C", home, "log", "-1", "--format=%H %cI"], capture_output=True, text=True).stdout.strip()
    if not rev:  # a release wheel has no checkout; name the installed package instead
        try:
            from importlib.metadata import version

            rev = f"ttnn {version('ttnn')} (release wheel)"
        except Exception:  # noqa: BLE001
            rev = ""
    device = ttnn.open_device(device_id=0)
    try:
        print(f"arch {device.arch()}   tt-metal {rev or 'unknown'}   q/k/v bfloat16 TILE, B=1 H=1 D=64")
        for T in (449, 450, 448):
            torch.manual_seed(0)
            q, k, v = torch.randn(1, 1, T, 64) * 0.3, torch.randn(1, 1, T, 64), torch.randn(1, 1, T, 64)
            ref = torch.softmax(q @ k.transpose(-1, -2), -1) @ v
            for kpad, vpad in ((0.0, 0.0), (1e38, 0.0), (0.0, 1e38), (1e38, 1e38)):
                qt, kt, vt = upload(q, T, 0.0, device), upload(k, T, kpad, device), upload(v, T, vpad, device)
                o = ttnn.to_torch(ttnn.transformer.scaled_dot_product_attention(
                    qt, kt, vt, is_causal=False, scale=1.0)).float()
                # Same-precision control: the same bfloat16 tensors, padding included, through
                # matmul -> softmax -> matmul on device.
                e = ttnn.to_torch(ttnn.matmul(ttnn.softmax(ttnn.matmul(qt, kt, transpose_b=True), dim=-1), vt)).float()
                print(f"T={T} (T%32={T % 32:>2})  K pad {kpad:<6g} V pad {vpad:<6g}  "
                      f"SDPA {err(o, ref)} | matmul-softmax-matmul {err(e, ref)}")
    finally:
        ttnn.close_device(device)


if __name__ == "__main__":
    main()
```

</details>

<details>
<summary><code>probe_sdpa_padding.py</code> (the HiFi4 / fp32 sweep)</summary>

```python
# SPDX-FileCopyrightText: (c) 2026 Tenstorrent AI ULC
#
# SPDX-License-Identifier: Apache-2.0
"""SDPA and the contents of tile padding: a model-free reproduction.

The CosyVoice flow estimator calls `ttnn.transformer.scaled_dot_product_attention` unmasked
and non-causal, on q/k/v whose sequence length is not a tile multiple. Captured from the
model, those tensors carry non-finite values in their padding rows, and at some lengths
SDPA then returns a wrong answer (PCC 0.01) where the matmul-softmax-matmul chain on the
same tensors is exact. Replaying the captured tensors with the padding zeroed fixes it.

This probe builds the case from scratch: random q/k/v with the logical rows fixed, then a
chosen value written into the padding rows of chosen operands -- uploaded at the padded
length and reshaped to the logical one with `skip_padding_fill=True`, so the values really
are in the padding on device. Each row of the output is one (T, pattern) pair, scored
against torch on the logical rows. The question is whether the damage depends on the
length, on the value, on the operand, or on the combination.

    python3 probe_sdpa_padding.py [--lengths 417,448,449,450] [--values big,inf,-inf,nan] [--operands q,k,v,kv]
"""
from __future__ import annotations

import argparse
import json
import sys

import torch

import ttnn


def pcc(a, b):
    a, b = a.flatten().double(), b.flatten().double()
    ok = torch.isfinite(a) & torch.isfinite(b)
    if ok.sum() < 2:
        return float("nan")
    a, b = a[ok] - a[ok].mean(), b[ok] - b[ok].mean()
    den = float(a.norm() * b.norm())
    return float((a * b).sum()) / den if den > 0 else float("nan")


VALUES = {"zero": 0.0, "big": 1.0e38, "inf": float("inf"), "-inf": float("-inf"), "nan": float("nan")}


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--lengths", default="385,417,433,448,449,450,451,465,481,897")
    ap.add_argument("--values", default="big,inf,nan")
    ap.add_argument("--operands", default="k,v,kv,qk,qv,qkv")
    ap.add_argument("--rows", default="all", help="which padding rows get the value: all | first | last")
    ap.add_argument("--out", default=None)
    args = ap.parse_args()

    device = ttnn.open_device(device_id=0, l1_small_size=32768)
    cc = ttnn.init_device_compute_kernel_config(
        device.arch(), math_fidelity=ttnn.MathFidelity.HiFi4, math_approx_mode=False,
        fp32_dest_acc_en=True, packer_l1_acc=True)
    rows = []
    try:
        print(f"arch {device.arch()}   B=2 h=8 d=64, non-causal, no mask, scale=1   padding rows: {args.rows}")
        print(f"{'T':>5} {'T%32':>4} {'value':>6} {'operands':>8} | {'pcc':>9} {'nonfin':>7} {'max|out|':>10}")
        for T in [int(x) for x in args.lengths.split(",")]:
            Tp = -(-T // 32) * 32
            g = torch.Generator().manual_seed(T)
            base = {n: torch.randn(2, 8, T, 64, generator=g) * s for n, s in (("q", 0.3), ("k", 1.0), ("v", 1.0))}
            ref = torch.softmax(base["q"].double() @ base["k"].double().transpose(-1, -2), -1) @ base["v"].double()

            def upload(name, value, targets):
                p = torch.zeros(2, 8, Tp, 64)
                p[..., :T, :] = base[name]
                if name in targets and Tp > T:
                    sel = {"all": slice(T, Tp), "first": slice(T, T + 1), "last": slice(Tp - 1, Tp)}[args.rows]
                    p[..., sel, :] = value
                t = ttnn.from_torch(p, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
                if Tp == T:
                    return t
                return ttnn.reshape(t, ttnn.Shape([2, 8, T, 64]), ttnn.Shape([2, 8, Tp, 64]), skip_padding_fill=True)

            patterns = [("zero", "-")] + [(v, o) for v in args.values.split(",") for o in args.operands.split(",")]
            for vname, ops in patterns:
                val = VALUES[vname]
                q, k, v = (upload(n, val, set(ops) if ops != "-" else set()) for n in ("q", "k", "v"))
                o = ttnn.to_torch(ttnn.transformer.scaled_dot_product_attention(
                    q, k, v, is_causal=False, scale=1.0, compute_kernel_config=cc)).float()
                r = {"T": T, "value": vname, "operands": ops, "pcc": pcc(o, ref),
                     "nonfinite": int((~torch.isfinite(o)).sum()),
                     "max": float(o[torch.isfinite(o)].abs().max()) if torch.isfinite(o).any() else float("nan")}
                rows.append(r)
                bad = r["nonfinite"] > 0 or not (r["pcc"] > 0.999)
                print(f"{T:>5} {T % 32:>4} {vname:>6} {ops:>8} | {r['pcc']:>9.5f} {r['nonfinite']:>7} {r['max']:>10.4g}"
                      f"{'   <-- WRONG' if bad else ''}", flush=True)
                for t in (q, k, v):
                    ttnn.deallocate(t)
    finally:
        if args.out:
            with open(args.out, "w") as fh:
                json.dump(rows, fh, indent=1)
        ttnn.close_device(device)
    print("SDPA_PADDING_DONE")
    return 0


if __name__ == "__main__":
    sys.exit(main())
```

</details>
